src/Application/Actions/Donations/Update.php

```diff
@@ -124,9 +126,23 @@ protected function action(): Response
             );
         }
 
-        $response = $this->addData($donation, $donationData);
-
-        return $response;
+        $retryCount = 0;
+        while ($retryCount < self::MAX_UPDATE_RETRY_COUNT) {
+            try {
+                if ($retryCount > 0) {
+                    $this->entityManager->refresh($donation, LockMode::PESSIMISTIC_WRITE);
+                }
+                return $this->addData($donation, $donationData);
+            } catch (LockWaitTimeoutException $lockWaitTimeoutException) {
+                \usleep(100_000 * (2 ** $retryCount)); // pause for 0.1, 0.2, 0.4 and then 0.8s before giving up.
+                $retryCount++;
+            }
+        }
+        throw new \Exception(
+            'Retry count exceeded trying to update donation, retried ' . self::MAX_UPDATE_RETRY_COUNT . " times",
```

Comment:
@bdsl I think it'd be nice to mention the donation id in this exception message, so we as devs don't have to dig deeper into logs to find the affected donation id.

Also, I think you should be logging the `$retryCount` instead of `self::MAX_UPDATE_RETRY_COUNT`. It's much more truthful. Imagine we introduce a bug without knowing that caused the retries to happen 10 times where `self::MAX_UPDATE_RETRY_COUNT` was still 4. In such a case, the exception would incorrectly log we retried 4 times, when in fact the truth would be 10 times. The incorrect log in such an instance would add much more confusion when trying to debug what happened wrong, let alone noticing the issue in the first place.

Reply:
Two good ideas.

```suggestion
            "Retry count exceeded trying to update donation #{$donation->getId()} , retried $retryCount times",
```

Reply:
Nice one @bdsl!